### PR TITLE
kv: don't leak raft application tracing spans on or after ErrRemoved

### LINKED ISSUE #53677: kv: raft application leaks traces on ErrRemoved
Noticed while investigating https://github.com/cockroachdb/cockroach/issues/53403.

Screen Shot 2020-08-28 at 4 53 06 PM

In a cluster with net/trace enabled for long periods of time, we begin to see an accumulation of leaked tracing spans. This appears to be fallout from https://github.com/cockroachdb/cockroach/commit/c4adb7a9cd09ec96b9c4d8acec35a2ece2b9a880, so it affects v20.1 and v20.2. That commit introduced tracing for raft application of non-local proposals. This means that we need to properly clean up (`FinishAndAckOutcome`) these commands even in the presence of replica removal. We don't currently do so.

This currently goes wrong in two ways. First, we don't clean up the command that caused an `ErrRemoved`. Second, we don't clean up any commands after this command in the same batch. This second issue may be even worse than a tracing leak – it may be possible for local proposals to be later in the same batch, in which case we would leak them entirely and leave them hanging indefinitely. It's not clear whether this second hazard is possible in practice.

### PR DESCRIPTION
**Pull request by nvb**
Fixes #53677.

This change ensures that we properly finish tracing spans of Raft commands that throw ErrRemoved errors in ApplySideEffects. It then ensures that we properly finish tracing spans of Raft commands that follow a command that throws an `ErrRemoved`. Before this, these commands would be abandoned and would never be finished. The effects of this are theoretically even worse than those fixed in the previous commit because these leaked commands could be locally proposed, so we may be abandoning a local proposer indefinitely.

It's not clear that we ever saw an instance of this. It seems rare for a local proposal to end up in the same CommittedEntries batch as a command that removes a replica because of the lease requirements, but it doesn't seem impossible, especially of the local proposal was a RequestLease request.

I was originally intending to do something more dramatic and make `replicaStateMachine.ApplySideEffects` responsible for acknowledging proposers in all cases, but doing so turned out to be pretty invasive so I was concerned that it would be harder to backport to v20.2 and to v20.1. I may revisit that in the future.

Release justification: low risk, high benefit changes to existing functionality.

### CODE REVIEW

**Review by ajwerner — APPROVED** [review]
Defeat the linter and :lgtm_strong: 

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54140#-:-MGooO7Z4OyyFZmMmD_d:bi3oxcg)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/kv/kvserver/apply/cmd.go, line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54140#-MGonMde9iDtCHTzTMYR:-MGonMde9iDtCHTzTMYS:b-ru34d8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4ce989abb5b155c477de425cf410cf5a77ecd9d/pkg/kv/kvserver/apply/cmd.go#L208)):*
> ```Go
> 
> // In the following three functions, fn is written with ctx as a 2nd param
> // because callers wants to bind it to methods that has Commands (or variants)
> ```

“callers wants”

This comment is generally hard to parse

**Review by nvb — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54140#-:-MGuj1nI0N_jQFvKaWEG:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/kv/kvserver/apply/cmd.go, line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/54140#-MGonMde9iDtCHTzTMYR:-MGuixbLD0RCgBOm-jSG:b-1szjoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4ce989abb5b155c477de425cf410cf5a77ecd9d/pkg/kv/kvserver/apply/cmd.go#L208)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

“callers wants”

This comment is generally hard to parse 
</blockquote></details>

Done. Cleaned it up a bit too.
